openhtf/output/web_gui/__init__.py

```diff
@@ -223,26 +223,33 @@ def __getitem__(self, hostport):  # pylint:disable=invalid-name
 
   def _discover(self):
     """Discover stations through the station API."""
+    hostports = []
     for station in station_api.Station.discover():
       hostport = Hostport(station.host, station.station_api_port)
       self.stations.setdefault(hostport, station)
+      hostports.append(hostport)
 
+    self._handle_stations(hostports)
+
+  def _handle_stations(self, hostports):
+    for hostport in hostports:
       try:
         self.watch_tests(hostport)
       except station_api.StationUnreachableError:
         _LOG.debug('Station at %s is unreachable.', hostport)
-
     if self._on_discovery_callback:
       self._on_discovery_callback(self.stations)
 
   def run(self):
     """Continuously scan for new stations and add them to the store."""
-    if self._disable_discovery:
-      _LOG.debug("Station discovery is disabled; StationStore won't update.")
-      return
+    self._handle_stations(self.stations.iterkeys())
     while not self._stop_event.is_set():
-      self._discover()
-      self._stop_event.wait(self._discovery_interval_s)
+      if self._disable_discovery:
+        _LOG.debug('Station discovery is disabled; using static station info')
+        self._handle_stations(self.stations.iterkeys())
```

Comment:
Seems like it's reasonable for `_disable_discovery` to be `True` and `conf.stations` to be set as well, in which case, we want to handle the stations in `self.stations` even when `_disable_discovery` is False.

If that sounds right, then let's remove the call to `_handle_stations` from `_discover`. Then in this loop, we can call `_discover`, followed by a call to `self._handle_stations(self.stations.iterkeys())` to get all the stations.